Crash when adding static-modifier

When compiling the following list code

```
datatype List<T> = Nil | Cons(T, List<T>) {
  function method App(ys: List<T>): List<T> {
    match this
      case Nil => ys
      case Cons(x, xs) => Cons(x, xs.App(ys))
  }
  
  static function method Concat<T>(l: List<List<T>>): List<T> {
    match l
      case Nil => Nil
      case Cons(x, xs) => x.App(Concat(xs))
  }

  lemma AppAssoc(m: List<T>, n: List<T>)
    ensures App(m).App(n) == App(m.App(n))
  {}

  static lemma ConcatApp<T>(l1: List<List<T>>, l2: List<List<T>>)
    ensures Concat(l1.App(l2)) == Concat(l1).App(Concat(l2))
  {
      match l1
        case Nil =>
        case Cons(x, xs) =>
          x.AppAssoc(Concat(xs), Concat(l2));
  }
}
```

Dafny 2.3.0.10506 crashes with

```
[ERROR] FATAL UNHANDLED EXCEPTION: System.InvalidCastException: Specified cast is not valid.
  at Microsoft.Dafny.Translator.RecursiveCallParameters (Microsoft.Boogie.IToken tok, Microsoft.Dafny.MemberDecl member, System.Collections.Generic.List`1[T] typeParams, System.Collections.Generic.List`1[T] ins, System.Collections.Generic.Dictionary`2[TKey,TValue] substMap, Microsoft.Dafny.Expression& receiver, System.Collections.Generic.List`1[Microsoft.Dafny.Expression]& arguments) [0x0000c] in <31773b6dd1e3472886c3b658c150744f>:0
  at Microsoft.Dafny.Translator.AddMethodImpl (Microsoft.Dafny.Method m, Microsoft.Boogie.Procedure proc, System.Boolean wellformednessProc) [0x00478] in <31773b6dd1e3472886c3b658c150744f>:0
  at Microsoft.Dafny.Translator.AddMethod_Top (Microsoft.Dafny.Method m) [0x00128] in <31773b6dd1e3472886c3b658c150744f>:0
  at Microsoft.Dafny.Translator.AddClassMembers (Microsoft.Dafny.TopLevelDeclWithMembers c, System.Boolean includeMethods) [0x003d0] in <31773b6dd1e3472886c3b658c150744f>:0
  at Microsoft.Dafny.Translator.AddTypeDecl (Microsoft.Dafny.RevealableTypeDecl d) [0x000ca] in <31773b6dd1e3472886c3b658c150744f>:0
  at Microsoft.Dafny.Translator.DoTranslation (Microsoft.Dafny.Program p, Microsoft.Dafny.ModuleDefinition forModule) [0x00437] in <31773b6dd1e3472886c3b658c150744f>:0
  at Microsoft.Dafny.Translator+<Translate>d__47.MoveNext () [0x000ea] in <31773b6dd1e3472886c3b658c150744f>:0
  at Microsoft.Dafny.DafnyDriver+<Translate>d__7.MoveNext () [0x00119] in <81519176a5d341bb8a45e2d5f929b2fa>:0
  at Microsoft.Dafny.DafnyDriver.Boogie (System.String baseName, System.Collections.Generic.IEnumerable`1[T] boogiePrograms, System.String programId, System.Collections.Generic.Dictionary`2[System.String,Microsoft.Boogie.PipelineStatistics]& statss, Microsoft.Boogie.PipelineOutcome& oc) [0x0012f] in <81519176a5d341bb8a45e2d5f929b2fa>:0
  at Microsoft.Dafny.DafnyDriver.ProcessFiles (System.Collections.Generic.IList`1[T] dafnyFiles, System.Collections.ObjectModel.ReadOnlyCollection`1[T] otherFileNames, Microsoft.Dafny.ErrorReporter reporter, System.Boolean lookForSnapshots, System.String programId) [0x00222] in <81519176a5d341bb8a45e2d5f929b2fa>:0
  at Microsoft.Dafny.DafnyDriver.ThreadMain (System.String[] args) [0x0004b] in <81519176a5d341bb8a45e2d5f929b2fa>:0
  at Microsoft.Dafny.DafnyDriver+<>c__DisplayClass1_0.<Main>b__0 () [0x00001] in <81519176a5d341bb8a45e2d5f929b2fa>:0
  at System.Threading.ThreadHelper.ThreadStart_Context (System.Object state) [0x00014] in <55adae4546cd485ba70e2948332ebe8c>:0
  at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00071] in <55adae4546cd485ba70e2948332ebe8c>:0
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00000] in <55adae4546cd485ba70e2948332ebe8c>:0
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state) [0x0002b] in <55adae4546cd485ba70e2948332ebe8c>:0
  at System.Threading.ThreadHelper.ThreadStart () [0x00008] in <55adae4546cd485ba70e2948332ebe8c>:0
```

--- a/Source/Dafny/Translator.cs
+++ b/Source/Dafny/Translator.cs
@@ -3646,7 +3646,7 @@ namespace Microsoft.Dafny {
       Contract.Ensures(Contract.ValueAtReturn(out arguments) != null);
 
       if (member.IsStatic) {
-        receiver = new StaticReceiverExpr(tok, (ClassDecl)member.EnclosingClass, true);  // this also resolves it
+        receiver = new StaticReceiverExpr(tok, (TopLevelDeclWithMembers)member.EnclosingClass, true);  // this also resolves it
       } else {
         receiver = new ImplicitThisExpr(tok);
         receiver.Type = Resolver.GetReceiverType(tok, member);  // resolve here

--- a/Test/git-issues/git-issue-784.dfy
+++ b/Test/git-issues/git-issue-784.dfy
@@ -1,0 +1,33 @@
+// RUN: %dafny /compile:0 "%s" > "%t"
+// RUN: %dafny /noVerify /compile:4 /compileTarget:cs "%s" >> "%t"
+// RUN: %dafny /noVerify /compile:4 /compileTarget:js "%s" >> "%t"
+// RUN: %dafny /noVerify /compile:4 /compileTarget:go "%s" >> "%t"
+// RUN: %dafny /noVerify /compile:4 /compileTarget:java "%s" >> "%t"
+// RUN: %diff "%s.expect" "%t"
+
+datatype List<T> = Nil | Cons(T, List<T>) {
+  function method App(ys: List<T>): List<T> {
+    match this
+      case Nil => ys
+      case Cons(x, xs) => Cons(x, xs.App(ys))
+  }
+  
+  static function method Concat<T>(l: List<List<T>>): List<T> {
+    match l
+      case Nil => Nil
+      case Cons(x, xs) => x.App(Concat(xs))
+  }
+
+  lemma AppAssoc(m: List<T>, n: List<T>)
+    ensures App(m).App(n) == App(m.App(n))
+  {}
+
+  static lemma ConcatApp<T>(l1: List<List<T>>, l2: List<List<T>>)
+    ensures Concat(l1.App(l2)) == Concat(l1).App(Concat(l2))
+  {
+      match l1
+        case Nil =>
+        case Cons(x, xs) =>
+          x.AppAssoc(Concat(xs), Concat(l2));
+  }
+}

--- a/Test/git-issues/git-issue-784.dfy.expect
+++ b/Test/git-issues/git-issue-784.dfy.expect
@@ -1,0 +1,11 @@
+
+Dafny program verifier finished with 4 verified, 0 errors
+
+Dafny program verifier finished with 0 verified, 0 errors
+
+Dafny program verifier finished with 0 verified, 0 errors
+
+Dafny program verifier finished with 0 verified, 0 errors
+
+Dafny program verifier finished with 0 verified, 0 errors
+File git-issue-784/git_issue_784.java contains the partially compiled program

